tls/s2n_recv.c

```diff
@@ -204,6 +204,15 @@ ssize_t s2n_recv(struct s2n_connection * conn, void *buf, ssize_t size, s2n_bloc
     return bytes_read;
 }
 
+ssize_t s2n_recv(struct s2n_connection * conn, void *buf, ssize_t size, s2n_blocked_status * blocked)
+{
+    ENSURE_POSIX(!conn->recv_lock, S2N_ERR_CONCURRENT_CALLS);
+    conn->recv_lock = true;
```

Comment:
This will exhibit a time-of-check-to-time-of-use bug ([TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use); see [linearizability](https://en.wikipedia.org/wiki/Linearizability)) when called from multiple threads. Without an atomic compare-and-swap, the best we can claim is we detect reentrancy issues.

In fact this brings up another issue: [In our docs](https://github.com/awslabs/s2n/blob/main/docs/USAGE-GUIDE.md#s2n_connection_new) we make the claim that a connection may be used in two threads:

> These connection objects can be operated on by up to two threads at a time, one sender and one receiver, but neither sending nor receiving are atomic, so if these objects are being called by multiple sender or receiver threads, you must perform your own locking to ensure that only one sender or receiver is active at a time.

I'm skeptical of this claim. There is a certain amount of shared state between the two sides and we do not have any kind of synchronization ensuring linearizability. This is actually a really good example of potential issues that arise. Both the `recv_lock` and `send_lock` are stored at the same byte in memory. As we can see in the [output assembly](https://godbolt.org/z/joav1r) we end up using an `or` and `and` to both set and unset these flags. I don't believe these are guaranteed to be atomic operations by the CPU, at least not across architectures.

In short, we should remove that statement from our docs as well :)

Reply:
To be clear, are you suggesting that a customer can't call send from one thread while calling recv from another thread? Because I don't think the two calls actually share any mutable state. I believe calling both at the same time is possible.

Reply:
Ah although I'm realizing we might not have kept this in mind while adding post-handshake messages. I'll create an issue for that.

Reply:
> I don't think the two calls actually share any mutable state

These flags are an example of us sharing unsynchronized mutable state. Flags are padded to at least the nearest byte and both of these flags share that byte. Since each of these methods ends up setting and clearing the flags with the non-atomic `or` and `and` operations, respectively, there is potential for one thread calling `s2n_send` which will clobber the expected state of `s2n_recv`.

I would not be surprised if we have other issues like this. The first thing that comes to mind is synchronizing key updates between the two halves.

Reply:
Understood. That can probably be solved by not making these particular values bitflags.

We shouldn't remove that line from the docs. We need to be able to support send and receive on different threads. We should address any issues that would make that line not true. Ideally we'd also have a reliable way to test for these problems :\ I'll create an issue to investigate KeyUpdate, because I think you're right about that being a problem (I think we're clearing the "key_update_pending" flag too early).

Reply:
Looks like the separate `bool`s should fix it: https://godbolt.org/z/dYzd98